### ISSUE TITLE
Improvements to fuse optimization

### ISSUE THREAD

**Issue by sanderland** [pytorch/pytorch#87777]
### 🚀 The feature, motivation and pitch

Some optimization packages call pytorch's fuse function to try optimization without knowing much about the model, notably ipex and packages which call ipex.
This operation relies on symbolic tracing, which fails frequently due to the limitations on conditional statements.

A possible improvement here is before tracing, filter the patterns by those that could work given the submodules, i.e. if there is no conv or batch norm, there is no reason to include the pattern. If there are none left after this step, just return the model.

### Alternatives

Alternatively or additionally, symbolic tracing (or it's use in fuse) could be improved.
The points of failure in symbolic tracing are outlined in the ipex issue and are mainly:

* input checking on optional arguments, which could be resolved by allowing concrete_args in fuse
* dimension checks, which could be improved by allowing a proxy to have a dimension or even a shape (although this goes quite far, and I can see the potential for some strange bugs here)

### Additional context

[this issue](https://github.com/intel/intel-extension-for-pytorch/issues/250) contains some examples of typical failures.

**Comment by vkuzo** [comment]
Hi @sanderland , these are reasonable improvements.  cc @jerryzh168  and @andrewor14 , any opposition to exposing `concrete_args` in `prepare_fx/fuse_fx`?

High level, PyTorch quantization is planning to onboarding to the new torchdynamo export-only frontend (https://github.com/pytorch/torchdynamo/blob/main/documentation/FAQ.md#do-i-still-need-to-export-whole-graphs) in the future, so we are unlikely to get to this issue ourselves in the near future.  We would be happy to accept a PR if someone wants to get to it sooner!

**Comment by jerryzh168** [comment]
> any opposition to exposing concrete_args in prepare_fx/fuse_fx?

sounds good to me
